### PR TITLE
Adapt drift compensation code to the way Flow checks convergence. 

### PR DESCRIPTION
**Pull request by totto82**
Depends on https://github.com/OPM/opm-models/pull/699

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this opm-models=699 please

**Comment by bska** [comment]
> jenkins build this opm-models=699

I think I've said before that we really should trigger multi-repo build checks from the most upstream repository&mdash;in this case opm-models.  Otherwise, we don't run all unit tests in the upstream repositories.

**Comment by akva2** [comment]
you have to rebase.

**Comment by totto82** [comment]
> you have to rebase. i removed that function in my refactor as it was unused, you'll have to reintroduce it.

I found a better work-around. Just asked the eclState directly.

File: ebos/eclproblem.hh [hunk 1]
```diff
@@ -2040,29 +2044,31 @@ public:
...
+        bool compensateDrift =  schedule.numWells(episodeIdx) > 0 && aquifer.active();
+        if (enableDriftCompensation_ && compensateDrift) {
             const auto& simulator = this->simulator();
             const auto& model = this->model();
 
             // we need a higher maxCompensation than the Newton tolerance because the
             // current time step might be shorter than the last one
-            Scalar maxCompensation = 10.0*model.newtonMethod().tolerance();
-
+            Scalar maxCompensation = model.newtonMethod().tolerance()/10;
```
Comment (blattms) [inline ebos/eclproblem.hh:2059]: Maybe we should change the comment accordingly?

File: ebos/eclproblem.hh [hunk 2]
```diff
@@ -3147,6 +3153,33 @@ private:
...
+            for (unsigned phaseIdx = 0; phaseIdx < numPhases; ++phaseIdx) {
+                if (!FluidSystem::phaseIsActive(phaseIdx))
+                    continue;
+
+                sumInvB[phaseIdx] += dofFluidState.invB(phaseIdx);
+            }
+        }
+        const auto& comm = this->simulator().vanguard().grid().comm();
+        comm.sum(sumInvB.data(),sumInvB.size());
+        Scalar numTotalDof = comm.sum(numDof);
```
Comment (blattms) [inline ebos/eclproblem.hh:3170]: I assume that these are all dofs regardless whether they are at ghost cells or not. Then both sumInvB and numTotalDof will be too high in parallel as dofs that are in the ghost area on a process will be counted twice.

The easiest fix would be to iterate over the interior of the grid. Not sure whether there is a way to get the corresponding dof indices then.

**Review by blattms — COMMENTED** [review]
Apart from the paralllel issues this looks good. They might not have such a big effect as long as the interior is big, but I think these need to fixed eventually. If we do it now we at least won't forget about it.

File: ebos/eclproblem.hh (hunk 2 above)
Comment (totto82) [reply ebos/eclproblem.hh:2978]: This should be fixed now

File: ebos/eclproblem.hh (hunk 1 above)
Comment (totto82) [reply ebos/eclproblem.hh:1856]: done

**Comment by totto82** [comment]
jenkins build this please

File: ebos/eclproblem.hh [hunk 3]
```diff
@@ -2943,6 +2950,44 @@ private:
...
+        std::vector<Scalar> sumInvB(numPhases, 0.0);
+        const auto& gridView = this->gridView();
+        ElementContext elemCtx(this->simulator());
+        auto elemIt = gridView.template begin</*codim=*/0>();
+        const auto& elemEndIt = gridView.template end</*codim=*/0>();
+        for (; elemIt != elemEndIt; ++elemIt) {
+            const auto& elem = *elemIt;
+            if (elem.partitionType() != Dune::InteriorEntity)
+                continue;
+
```
Comment (blattms) [inline ebos/eclproblem.hh:2963]: BTW There is a shortcut for this in modern DUNE that simplifies code:
```
for(const auto& elem: elements(gridView, Dune::Partitions::interior))

**Review by blattms — APPROVED** [review]
Up to you whether to use the simpler modern way.

File: ebos/eclproblem.hh (hunk 3 above)
Comment (totto82) [reply ebos/eclproblem.hh]: Thanks. Changed. Who doesn't want to be modern!

**Comment by blattms** [comment]
Thanks a lot. Will merge both of them.

**Comment by blattms** [comment]
I guess you need to do an update_data first because of the test failing expectedly, first.

**Pull request merged**